File: 2024/April/4-3 79 Word Search/WordSearch.py
from typing import List

class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        current = 0
        visited = [[0 for x in range(len(board))] for y in range(len(board[0]))]

        def search(i: int, j: int, word: str) -> bool:
            nonlocal current, visited, board

            if i < 0 or i >= len(board) or j < 0 or j >= len(board[0]):
                return False
            if board[i][i] != word[0]:
                return False
            if visited[i][j] == current:
                return False    
            if len(word) == 1:
                return True
            
            visited[i][j] = current
            nextWord = word[1:]

            res = search(i-1, j, nextWord) or search(i, j-1, nextWord) or search(i+1, j, nextWord) or search(i, j+1, nextWord)
            if res is False:
                visited[i][j] = current - 1
            
            return res

        for i in range(len(board)):
            for j in range(len(board[0])):
                current += 1
                if search(i, j, word):
                    return True
        
        return False

File: 2024/April/4-3 79 Word Search/WordSearch.java
class Solution {
    int[][] visited;
    char[][] board;
    int current;

    public boolean exist(char[][] board, String word) {
        visited = new int[board.length][board[0].length];
        this.board = board;
        current = 0;

        for(int i = 0; i < board.length; i++) {
            for(int j = 0; j < board[0].length; j++) {
                current++;
                if(search(i, j, word)) {
                    return true;
                }
            }
        }

        return false;
    }

    private boolean search(int i, int j, String word) {
        if(i < 0
        || i >= visited.length
        || j < 0
        || j >= visited[0].length) {
            return false;
        }

        if(board[i][j] != word.charAt(0)) {
            return false;
        }

        if(visited[i][j] == current) {
            return false;
        }

        if(word.length() == 1) {
            return true;
        }

        visited[i][j] = current;
        String nextWord = word.substring(1);

        boolean res = search(i-1, j, nextWord)
                   || search(i, j-1, nextWord)
                   || search(i+1, j, nextWord)
                   || search(i, j+1, nextWord); 
        if(res == false) {
            visited[i][j] = current - 1;
        }

        return res;
    }
}